src/main/java/de/hysky/skyblocker/skyblock/chocolatefactory/ChocolateFactorySolver.java

```diff
@@ -0,0 +1,222 @@
+				return List.of(ColorHighlight.green(rabbit.slot), ColorHighlight.yellow(bestRabbit.slot));
+			}
+		}
+
+		return List.of(ColorHighlight.yellow(bestRabbit.slot));
+	}
+
+	private static void updateFactoryInfo(Int2ObjectMap<ItemStack> slots) {
+		cpsIncreaseFactors.clear();
+
+		for (int i = 29; i <= 33; i++) { // The 5 rabbits slots are in 29, 30, 31, 32 and 33.
+			ItemStack item = slots.get(i);
+			if (item.isOf(Items.PLAYER_HEAD)) {
+				getRabbit(item, i).ifPresent(cpsIncreaseFactors::add);
+			}
+		}
+
+		//Coach is in slot 42
+		getCoach(slots.get(42)).ifPresent(cpsIncreaseFactors::add);
+		RegexUtils.getLongFromMatcher(CHOCOLATE_PATTERN.matcher(slots.get(13).getName().getString())).ifPresent(l -> totalChocolate = l);
+
+		//Cps item (cocoa bean) is in slot 45
+		String cpsItemLore = getConcatenatedLore(slots.get(45));
+		Matcher cpsMatcher = CPS_PATTERN.matcher(cpsItemLore);
+		RegexUtils.getDoubleFromMatcher(cpsMatcher).ifPresent(d -> totalCps = d);
+
+		Matcher multiplierMatcher = TOTAL_MULTIPLIER_PATTERN.matcher(cpsItemLore);
+		RegexUtils.getDoubleFromMatcher(multiplierMatcher, cpsMatcher.hasMatch() ? cpsMatcher.end() : 0).ifPresent(d -> totalCpsMultiplier = d);
+
+		//Compare cost/cpsIncrease rather than cpsIncrease/cost to avoid getting close to 0 and losing precision.
+		cpsIncreaseFactors.sort(Comparator.comparingDouble(rabbit -> rabbit.cost() / rabbit.cpsIncrease())); //Ascending order, lower = better
+	}
+
+	private static void handleTooltip(ItemStack stack, Item.TooltipContext tooltipContext, TooltipType tooltipType, List<Text> lines) {
+		if (!SkyblockerConfigManager.get().helpers.chocolateFactory.enableChocolateFactoryHelper) return;
+		if (!(MinecraftClient.getInstance().currentScreen instanceof GenericContainerScreen screen) || !screen.getTitle().getString().equals("Chocolate Factory") ) return;
+
+		String lore = concatenateLore(lines);
+		Matcher costMatcher = COST_PATTERN.matcher(lore);
+		OptionalLong cost = RegexUtils.getLongFromMatcher(costMatcher);
+		if (cost.isEmpty() || totalChocolate == -1L || totalCps == -1.0) return;
+
+		lines.add(ItemTooltip.createSmoothLine());
+
+		lines.add(Text.literal("")
+		              .append(Text.literal("Time until upgrade: ").formatted(Formatting.GRAY))
+		              .append(formatTime((cost.getAsLong() - totalChocolate) / totalCps)));
+
+		if (cpsIncreaseFactors.isEmpty()) return;
+
+		for (Rabbit rabbit : cpsIncreaseFactors) {
+			if (rabbit.itemStack != stack) continue;
+
+			lines.add(Text.literal("")
+			              .append(Text.literal("CPS Increase: ").formatted(Formatting.GRAY))
+			              .append(Text.literal(DECIMAL_FORMAT.format(rabbit.cpsIncrease)).formatted(Formatting.GOLD)));
+
+			lines.add(Text.literal("")
+			              .append(Text.literal("Cost per CPS: ").formatted(Formatting.GRAY))
+			              .append(Text.literal(DECIMAL_FORMAT.format(rabbit.cost / rabbit.cpsIncrease)).formatted(Formatting.GOLD)));
```

Comment:
```suggestion
			              .append(Text.literal(String.format(Locale.ENGLISH, "%1$,.1f", rabbit.cost / rabbit.cpsIncrease)).formatted(Formatting.GOLD)));
```
use the same number format method as the rest of skyblocker (part 3)